File: 036. Validate BST/ValidateBst.java
class Program {
    // O(n) time | 0(d) space
    public static boolean validateBst(BST tree) {
        return validateBst(tree, Integer.MIN_VALUE, Integer.MAX_VALUE);
    }

    public static boolean validateBst(BST tree, int minValue, int maxValue) {
        if (tree.value < minValue || tree.value >= maxValue)
            return false;
        if (tree.left != null && !validateBst(tree.left, minValue, tree.value))
            return false;
        if (tree.right != null && !validateBst(tree.right, tree.value, maxValue))
            return false;
        return true;
    }

    static class BST {
        public int value;
        public BST left;
        public BST right;

        public BST(int value) {
            this.value = value;
        }
    }
}

File: 036. Validate BST/ValidateBst.kt
package com.algoexpert.program

open class BST(value: Int) {
    var value = value
    var left: BST? = null
    var right: BST? = null
}

// O(n) time | O(d) space
fun validateBst(tree: BST): Boolean {
    return validateBstHelper(tree, Int.MIN_VALUE, Int.MAX_VALUE)
}

fun validateBstHelper(tree: BST?, minValue: Int, maxValue: Int): Boolean {
    if (tree == null) return true
    if (tree.value < minValue || tree.value >= maxValue) return false
    val leftIsValid = validateBstHelper(tree.left, minValue, tree.value)
    return leftIsValid && validateBstHelper(tree.right, tree.value, maxValue)
}
